File: app/src/main/java/de/karzek/diettracker/presentation/main/settings/dialog/editAllergen/EditAllergensDialog.java
package de.karzek.diettracker.presentation.main.settings.dialog.editAllergen;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.FrameLayout;

import java.util.ArrayList;
import java.util.HashMap;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import de.karzek.diettracker.R;
import de.karzek.diettracker.presentation.TrackerApplication;
import de.karzek.diettracker.presentation.common.BaseDialog;
import de.karzek.diettracker.presentation.main.settings.dialog.editAllergen.adapter.EditAllergensListAdapter;
import de.karzek.diettracker.presentation.model.AllergenDisplayModel;

/**
 * Created by MarjanaKarzek on 07.06.2018.
 *
 * @author Marjana Karzek
 * @version 1.0
 * @date 07.06.2018
 */
public class EditAllergensDialog extends BaseDialog implements EditAllergensDialogContract.View {

    @Inject
    EditAllergensDialogContract.Presenter presenter;

    @BindView(R.id.recycler_view)
    RecyclerView recyclerView;
    @BindView(R.id.loading_view)
    FrameLayout loadingView;

    @BindView(R.id.dialog_action_dismiss)
    Button dismiss;
    @BindView(R.id.dialog_action_save)
    Button save;

    private View view;

    private SaveAllergenSelectionDialogListener listener;

    @Override
    protected void setupDialogComponent() {
        TrackerApplication.get(getContext()).createEditAllergensDialogComponent().inject(this);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.dialog_edit_checkbox_list, container, false);
        ButterKnife.bind(this, view);

        setupRecyclerView();

        dismiss.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dismiss();
                listener = null;
            }
        });

        save.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dismiss();
                presenter.saveAllergenSelection();
                listener.updateAllergens();
                listener = null;
            }
        });

        presenter.setView(this);
        presenter.start();

        return view;
    }

    private void setupRecyclerView() {
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(new EditAllergensListAdapter(presenter));
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        try {
            listener = (SaveAllergenSelectionDialogListener) getParentFragment();
        } catch (ClassCastException e) {
            throw new ClassCastException("fragment must implement SaveAllergenSelectionDialogListener");
        }
    }

    @Override
    public void updateRecyclerView(ArrayList<AllergenDisplayModel> allergens, HashMap<Integer, Boolean> allergenStatus) {
        ((EditAllergensListAdapter) recyclerView.getAdapter()).setAllergens(allergens, allergenStatus);
    }

    @Override
    public void showLoading() {
        loadingView.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideLoading() {
        loadingView.setVisibility(View.GONE);
    }

    @OnClick(R.id.reset_selection)
    void onResetSelectionClicked() {
        presenter.onResetSelectionClicked();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenter.finish();
        TrackerApplication.get(getContext()).releaseEditAllergensDialogComponent();
    }

    public interface SaveAllergenSelectionDialogListener {
        void updateAllergens();
    }
}

File: app/src/main/java/de/karzek/diettracker/presentation/main/settings/dialog/editAllergen/EditAllergensDialog.kt
package de.karzek.diettracker.presentation.main.settings.dialog.editAllergen

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.LinearLayoutManager
import de.karzek.diettracker.R
import de.karzek.diettracker.presentation.TrackerApplication
import de.karzek.diettracker.presentation.common.BaseDialog
import de.karzek.diettracker.presentation.main.settings.dialog.editAllergen.adapter.EditAllergensListAdapter
import de.karzek.diettracker.presentation.model.AllergenDisplayModel
import kotlinx.android.synthetic.main.dialog_edit_checkbox_list.*
import kotlinx.android.synthetic.main.snippet_loading_view.*
import java.util.*
import javax.inject.Inject

/**
 * Created by MarjanaKarzek on 07.06.2018.
 *
 * @author Marjana Karzek
 * @version 1.0
 * @date 07.06.2018
 */
class EditAllergensDialog : BaseDialog(), EditAllergensDialogContract.View {

    @Inject
    lateinit var presenter: EditAllergensDialogContract.Presenter

    private var currentView: View? = null

    private var listener: SaveAllergenSelectionDialogListener? = null

    override fun setupDialogComponent() {
        TrackerApplication.get(context!!).createEditAllergensDialogComponent().inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        currentView = inflater.inflate(R.layout.dialog_edit_checkbox_list, container, false)

        setupRecyclerView()

        dialog_action_dismiss.setOnClickListener {
            dismiss()
            listener = null
        }

        dialog_action_save.setOnClickListener {
            dismiss()
            presenter.saveAllergenSelection()
            listener!!.updateAllergens()
            listener = null
        }

        reset_selection.setOnClickListener { presenter.onResetSelectionClicked() }

        presenter.setView(this)
        presenter.start()

        return currentView
    }

    private fun setupRecyclerView() {
        recycler_view.setHasFixedSize(true)
        recycler_view.layoutManager = LinearLayoutManager(context)
        recycler_view.adapter = EditAllergensListAdapter(presenter)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        try {
            listener = parentFragment as SaveAllergenSelectionDialogListener?
        } catch (e: ClassCastException) {
            throw ClassCastException("fragment must implement SaveAllergenSelectionDialogListener")
        }

    }

    override fun updateRecyclerView(allergens: ArrayList<AllergenDisplayModel>, allergenStatus: HashMap<Int, Boolean>) {
        (recycler_view.adapter as EditAllergensListAdapter).setAllergens(allergens, allergenStatus)
    }

    override fun showLoading() {
        loading_view.visibility = View.VISIBLE
    }

    override fun hideLoading() {
        loading_view.visibility = View.GONE
    }

    override fun onDestroy() {
        super.onDestroy()
        presenter.finish()
        TrackerApplication.get(context!!).releaseEditAllergensDialogComponent()
    }

    interface SaveAllergenSelectionDialogListener {
        fun updateAllergens()
    }
}
